crypto/README.sparse_array

```diff
@@ -36,19 +35,115 @@ height:
 
 The tree height is dynamically increased as needed based on additions.
 
-The nodes themselves are allocated in a sparse manner.  Only nodes which
-exist along a path from the root of the tree to an added leaf will be
-allocated.  Again, the complexity is hidden and nodes are allocated on an
-as needed basis.
+An empty tree is represented by a NULL root pointer.  Inserting a value at
+index 1 results in the allocation of a top level node full of null pointers
+except for the single pointer to the user's data:
+ 
+        +----+
+        |Root|
+        |Node|
+        +-+--+
+          |
+          |
+          |
+          v
+        +-+-+---+---+---+---+
+        | 0 | 1 | 2 |...|N-1|
+        |nil|   |nil|...|nil|
+        +---+-+-+---+---+---+
+              |
+              |
+              |
+              v
+            +-+--+
+            |User|
+            |Data|
+            +----+
+        Index 1
 
-Values can be removed from the sparse array by setting their index
-position to NULL.  The data structure does not attempt to reclaim nodes
-or reduce the height of the tree on removal.
 
-Accesses to elements in the sparse array take O(log n) time where n is
-the largest element.  The base of the logarithm is SA_BLOCK_MAX, so for
-moderately small indices, single level (constant time) access is achievable.
-Space usage is O(m) where m is the number of elements in the array.
+Inserting at element 2N+2 creates a new root node and pushes down the old root
```

Comment:
maybe insert element 2N+3 to make it more obvious?

Reply:
I changed it to 2N+1 and the first element to 0 (from 1).